### PR TITLE
feat: Add `cooldown` and `max_concurrency` to `SlashCommandGroup`

### LINKED ISSUE #1935: SlashCommandGroup.before_invoke
### Summary

before_invoke is not called in Slash Command Group

### Reproduction Steps

Same steps as if i were using Bot.goup()

*  Creating a group
*  Creating  a command
*  Instruct  the program what to do before the command should be invoked

### Minimal Reproducible Code

```python
channel = d.SlashCommandGroup(name = "channel", description = "Create something", default_member_permissions = d.Permissions(manage_channels = True))

    @channel.command()
    async def test(self, ctx:d.ApplicationContext, text = None):
        print(ctx.guild.categories)
        return

    @channel.before_invoke
    async def check_channels(self, ctx:d.ApplicationContext):

        print("test, before_invoke")
        
        return
```

### Expected Results

Expected it to print out "before_invoke " in the terminal, before the command invoked.

### Actual Results

Function not even called

### Intents

bans, guilds, emojis, typing, members, messages, prresence, message_content

### System Information

- Python v3.10.10-final                                                                                                
- py-cord v2.4.None-final                                                                                              
- aiohttp v3.8.4                                                                                                        
- system info: Windows 10 10.0.19045

### Checklist

- [X] I have searched the open issues for duplicates.
- [X] I have shown the entire traceback, if possible.
- [X] I have removed my token from display, if visible.

### Additional Context

For further details on what has been already tested visit discord  channel search for before  & after  invoke by krigjo25

The code above is tested, and is just a demo, to show @before_invoke  doesnt work. The actual code can be found in the authors responsories. if more code information is required, the bot code source can be found at the authors responsories called Discord /pymod

* `@after_invoke` works as expected

Case has been discussed in the discord channel, recently closed since the helper also might think its a library issue. 

Thanks for reading this issues, and for the assistance.

sincerely 
@krigjo25

### LINKED ISSUE #2057: Possible bug in cooldowns
It might be that cooldowns and before_invoke would never run if you don't have max concurrency set for slash commands

https://github.com/Pycord-Development/pycord/blob/a09e728f7ff2b0911571341cefc5545043d7851a/discord/commands/core.py#L297-L308

Reported by @Om1609

### PR DESCRIPTION
**Pull request by Om1609**
## Summary

Adds `cooldown` and `max_concurrency` to `SlashCommandGroup`. Since `SlashCommandGroup` cannot be created using a decorator, these are created manually and passed directly to the init.

Also fixes #1935 by cleaning up the `ApplicationCommand.prepare` logic. (PR based on discussion in #2057)

## Information

- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting,
      examples, ...).

## Checklist

- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [ ] I have updated the documentation to reflect the changes.
- [x] If `type: ignore` comments were used, a comment is also left explaining why.
- [x] I have updated the changelog to include these changes.

### CODE REVIEW

**Comment by Middledot** [comment]
On third thought, since cooldowns and max concurrency can't be used by groups or affect their subcommands, you don't need to implement them. For the confusing formatting, it accounts for the fact that the og pr (#691) removed inheritance of these features, so it just checks one to filter out group objs. Instead of reimplementing cooldowns and max concurrency, you could just reformat the code to something like this:
```py
if not getattr(self, "_max_concurrency", None):
    # For this application, context can be duck-typed as a Message
    await self._max_concurrency.acquire(ctx)  # type: ignore # ctx instead of non-existent message

if hasattr(self, "_buckets"):
    try:
        self._prepare_cooldowns(ctx)
        await self.call_before_hooks(ctx)
    except:
        if not getattr(self, "_max_concurrency", None):
            await self._max_concurrency.release(ctx)  # type: ignore # ctx instead of non-existent message
        raise
```

File: CHANGELOG.md [hunk 1]
```diff
@@ -60,6 +60,8 @@ These changes are available on the `master` branch, but have not yet been releas
   ([#2042](https://github.com/Pycord-Development/pycord/pull/2042))
 - Added `icon` and `unicode_emoji` to `Guild.create_role`.
   ([#2086](https://github.com/Pycord-Development/pycord/pull/2086))
+- Added `cooldown` and `max_concurrency` to`SlashCommandGroup`.
```
Comment (Middledot) [inline CHANGELOG.md:63]:
```suggestion
- Added `cooldown` and `max_concurrency` to `SlashCommandGroup`.
```

**Comment by Om1609** [comment]
> Since cooldowns and max concurrency can't be used by groups or affect their subcommands, you don't need to implement them. 

They actually do control the subcommands. By setting a cooldown/max_con on the group, each individual subcommand get a **shared** cooldown/max_con. 

So if you had set 1 token per minute as the cooldown on the group, after first cmd invoke, you can't invoke any other command from the parent group. This also works with subgroups.

So this is actually a valid working feature.

**Comment by Middledot** [comment]
> They actually do control the subcommands. By setting a cooldown/max_con on the group, each individual subcommand gets a **shared** cooldown/max_con.

I actually confused this and the Invokable pr, because I addressed this there. Prefixed group commands don't have the same behaviour. Their group checks and subcommand checks are independent from each other, so I say the same should be for app commands

**Comment by Om1609** [comment]
> I actually confused this and the Invokable pr, because I addressed this there. 

Oh how?

> Prefixed group commands don't have the same behaviour. Their group checks and subcommand checks are independent from each other, so I say the same should be for app commands

The individual cmd's cooldown and max concurrency still apply. You could have 1 per 1 min cooldown on group, while also having 1 per 5 mins on a certain subcommand. (placeholder values for example purposes)

Eg-
Let's say we have group with three commands `/dance list`, `/dance preview`, and `/dance now`. 
Let's say we set the cooldown on `/dance` group as 1 per 1 min. And we set the cooldown for `/dance now` as 1 per 5 mins.

So now you can invoke **only 1** of any of the subcommands in a minute. If the command invoked was `/dance now`, then you could invoke `/dance list` or `/dance preview` after 1 min, but you will need to wait **5 mins** to invoke `/dance now` again. 

I hope this makes sense. I should also re-test if the logic I mentioned above actually holds true.

**Comment by Middledot** [comment]
Oh my bad, I confused it again. What I was talking about were checks of parents being run by subcommands. Cooldowns etc are managed differently. What you're talking about is an interesting use case that I never realized, plus it's consistent with prefixed commands, so yeah, nevermind

File: discord/commands/core.py [hunk 2]
```diff
@@ -1119,6 +1118,8 @@ def __init__(
         description: str | None = None,
         guild_ids: list[int] | None = None,
         parent: SlashCommandGroup | None = None,
+        cooldown: Any = None,
+        max_concurrency: Any = None,
```
Comment (Middledot) [inline discord/commands/core.py:1122]:
```suggestion
        cooldown: CooldownMapping | None = None,
        max_concurrency: MaxConcurrency | None = None,
```
Also, you need to import them in TYPE_CHECKING

Comment (Om1609) [reply discord/commands/core.py:1122]: Oh right. I'll do this tomorrow

**Review by Middledot — APPROVED** [review]
Tested, lgtm!

**Review by Lulalaby — APPROVED** [review]
239612949-082a1733-f980-48b2-b6eb-f634e33bd496

**Pull request merged**